diesel_derives/src/as_changeset.rs

```diff
@@ -58,21 +40,58 @@ pub fn derive(item: syn::DeriveInput) -> Result<proc_macro2::TokenStream, Diagno
             .emit();
     }
 
-    Ok(wrap_in_dummy_mod(quote!(
-        use diesel::query_builder::AsChangeset;
-        use diesel::prelude::*;
+    let mut generate_borrowed_changeset = true;
 
-        impl #impl_generics AsChangeset for &'update #struct_name #ty_generics
-        #where_clause
-        {
-            type Target = #table_name::table;
-            type Changeset = <(#(#ref_changeset_ty,)*) as AsChangeset>::Changeset;
+    let mut direct_changeset_ty = Vec::with_capacity(fields_for_update.len());
+    let mut direct_changeset_expr = Vec::with_capacity(fields_for_update.len());
+    let mut ref_changeset_ty = Vec::with_capacity(fields_for_update.len());
+    let mut ref_changeset_expr = Vec::with_capacity(fields_for_update.len());
 
-            fn as_changeset(self) -> Self::Changeset {
-                (#(#ref_changeset_expr,)*).as_changeset()
-            }
+    for field in fields_for_update {
+        if let Some(ty) = field.ty_for_serialize()? {
+            direct_changeset_ty.push(field_changeset_ty_serialize_as(
+                field,
+                &table_name,
+                treat_none_as_null,
+                &ty,
+            ));
+            direct_changeset_expr.push(field_changeset_expr_serialize_as(
+                field,
+                &table_name,
+                treat_none_as_null,
+                &ty,
+            ));
+
+            generate_borrowed_changeset = false; // as soon as we hit one field with #[diesel(serialize_as)] there is no point in generating the impl of AsChangeset for borrowed structs
```

Comment:
Why not?

Reply:
The conversion(`Into`) takes ownership. It would ask for a `Clone` on `serialize_as` fields if we generate the borrowed implementation. Any thought?

Reply:
We could require `AsRef` instead of `Into` as a trait bound.

Reply:
So we would require **two** traits to be implemented: `Into` for owned `AsChangeset`, and `AsRef` for borrowed one?

Reply:
@pksunkara Let me just summarize the options here shortly:

* We could require only an impl like `Into<Ty> for &MyType`, but that wouldn't allow users to reuse allocations 
* We could require both `Into<Ty> for MyType` and `Into<Ty> for &MyType`, that would require the user to provide both impls as there is no way to detect which impls actually exist from inside of a custom derive
* We could require that `Ty: Clone`, which would potentially hide allocations.
* We could require only `Into<Ty> for MyType` and only generate `Insertable<…> for Struct` (and not the reference variant). This pushes potential users into the "right" direction to reuse allocations + makes potential clones explicit

At least I feel that the last option is that one most in line with rusts design ideals.

Reply:
There's no way to check whether trait is implemented in runtime, right? So we need user pass the flag in (maybe through `serialize_as`'s attribute)?

Oh, yes. It seems **`AsRef` doesn't work**:( we cannot returns an owned object's reference.

Reply:
> We could require both `Into<Ty> for MyType` and `Into<Ty> for &MyType`, that would require the user to provide both impls as there is no way to detect which impls actually exist from inside of a custom derive

You can't emit code based on which type implements which trait, yes. But you can emit code that is itself conditional on whether a trait is implemented. We can emit something like:

```rust
impl<'a> AsChangeset for &'a T where Foo: AsRef<Bar> {

}
```

That should allow us to call `foo.as_ref()` inside the impl block. If a user doesn't need this conversion and hence doesn't implement `AsRef`, things should still work, but they will get a compile error of unresolved trait bounds in case they try to pass a reference of `T` in.

I am not sure we need it though.

Reply:
@thomaseizinger Restricting the impl in that way is unfortunately not possible. Rustc is too clever here and complains that `Foo` does not implement `AsRef<Bar>`. See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d1f5277fc1baa7f5c98d530df73ae448) for an example.

Reply:
Bugger, thanks for clearing it up.